core/shared/src/main/scala/zio/Promise.scala

```diff
@@ -207,104 +174,96 @@ final class Promise[E, A] private (
   }
 
-  private[zio] val unsafe: UnsafeAPI =
-    new UnsafeAPI {
-      def completeWith(io: IO[E, A])(implicit unsafe: Unsafe): Boolean = {
-        var action: () => Boolean = null.asInstanceOf[() => Boolean]
-        var retry                 = true
-
-        while (retry) {
-          val oldState = state.get
-
-          val newState = oldState match {
-            case Pending(joiners) =>
-              action = () => { joiners.foreach(_(io)); true }
-
-              Done(io)
-
-            case _ =>
-              action = Promise.ConstFalse
-
-              oldState
-          }
-
-          retry = !state.compareAndSet(oldState, newState)
+  @deprecated("Kept for binary compatibility only. Do not use", "2.1.15")
+  private[zio] def state: AtomicReference[Promise.internal.State[E, A]] =
+    unsafe.asInstanceOf[AtomicReference[Promise.internal.State[E, A]]]
+  private[zio] val unsafe: UnsafeAPI = new AtomicReference(Promise.internal.State.empty[E, A]) with UnsafeAPI { state =>
+    def completeWith(io: IO[E, A])(implicit unsafe: Unsafe): Boolean = {
+      @annotation.tailrec
+      def loop(): Boolean =
+        state.get match {
+          case pending: Pending[?, ?] =>
+            if (state.compareAndSet(pending, Done(io))) {
+              pending.complete(io)
+              true
+            } else {
+              loop()
+            }
+          case _: Done[?, ?] => false
         }
+      loop()
+    }
 
-        action()
-      }
+    def die(e: Throwable)(implicit trace: Trace, unsafe: Unsafe): Boolean =
+      completeWith(ZIO.die(e))
 
-      def die(e: Throwable)(implicit trace: Trace, unsafe: Unsafe): Boolean =
-        completeWith(ZIO.die(e))
+    def done(io: IO[E, A])(implicit unsafe: Unsafe): Unit = completeWith(io)
 
-      def done(io: IO[E, A])(implicit unsafe: Unsafe): Unit = {
-        var retry: Boolean                 = true
-        var joiners: List[IO[E, A] => Any] = null
+    def fail(e: E)(implicit trace: Trace, unsafe: Unsafe): Boolean =
+      completeWith(ZIO.fail(e))
 
-        while (retry) {
-          val oldState = state.get
+    def failCause(e: Cause[E])(implicit trace: Trace, unsafe: Unsafe): Boolean =
+      completeWith(ZIO.failCause(e))
 
-          val newState = oldState match {
-            case Pending(js) =>
-              joiners = js
-              Done(io)
-            case _ => oldState
-          }
+    def interruptAs(fiberId: FiberId)(implicit trace: Trace, unsafe: Unsafe): Boolean =
+      completeWith(ZIO.interruptAs(fiberId))
 
-          retry = !state.compareAndSet(oldState, newState)
-        }
+    def isDone(implicit unsafe: Unsafe): Boolean =
+      state.get().isInstanceOf[Done[?, ?]]
 
-        if (joiners ne null) joiners.foreach(_(io))
+    def poll(implicit unsafe: Unsafe): Option[IO[E, A]] =
+      state.get() match {
+        case _: Pending[?, ?] => None
+        case Done(value)      => Some(value)
```

Comment:
We can probably optimize this a little bit more:
```suggestion
        case Done(value)      => Some(value)
        case _                => None
```